src/playdate/scoreboards.nim

```diff
@@ -0,0 +1,122 @@
+  # BoardsListCallback* = proc(boards: ptr PDBoardsList, errorMessage: string)
+  ScoresCallback* = proc(scores: PDScoresList, errorMessage: string)
+
+var privatePersonalBestCallbacks = newSeq[PersonalBestCallback]()
+var privateAddScoreCallback: AddScoreCallback
+var privateScoresCallbacks = newSeq[ScoresCallback]()
+
+proc newPDScoresList(boardID: string, lastUpdated: uint32, scores: seq[PDScore]): PDScoresList =
+  result.boardID = boardID
+  result.lastUpdated = lastUpdated
+  result.scores = scores
+
+proc invokePersonalBestCallback(score: PDScorePtr, errorMessage: ConstChar) {.cdecl, raises: [].} =
+  let callback = privatePersonalBestCallbacks.pop() # first in, first out
+  if score == nil and errorMessage == nil:
+    callback(PDScore(), "Playdate-nim: No personal best")
+    return
+
+  if score == nil:
+    callback(PDScore(), $errorMessage)
+    return
+    
+  let domainScore = PDScore(value: score.value.uint32, rank: score.rank.uint32, player: $score.player)
+  playdate.scoreboards.freeScore(score)
+  callback(domainScore, $errorMessage)
+
+proc invokeAddScoreCallback(score: PDScorePtr, errorMessage: ConstChar) {.cdecl, raises: [].} =
+  if errorMessage != nil:
+    privateAddScoreCallback(PDScore(value: 0, rank: 0, player: ""), $errorMessage)
+    return
+    
+  # todo create newPDSCore constructor
+  let domainScore = PDScore(value: score.value.uint32, rank: score.rank.uint32, player: $score.player)
+  playdate.scoreboards.freeScore(score)
+  privateAddScoreCallback(domainScore, $errorMessage)
+
+proc invokeScoresCallback(scoresList: PDScoresListPtr, errorMessage: ConstChar) {.cdecl, raises: [].} =
+  privateAccess(PlaydateScoreboards)
+  let callback = privateScoresCallbacks.pop() # first in, first out
+  if scoresList == nil and errorMessage == nil:
+    let domainObject = newPDScoresList(boardID = "", lastUpdated = 0, scores = @[])
+    callback(domainObject, "Playdate-nim: No scores")
+    return
+
+  if scoresList == nil:
+    let domainObject = newPDScoresList(boardID = "", lastUpdated = 0, scores = @[])
+    callback(domainObject, $errorMessage)
+    return
+
+  privateAccess(SDKArray)
+  let length = scoresList.count.cint
+  let cArray = SDKArray[PDScoreRaw](data: cast[ptr UncheckedArray[PDScoreRaw]](scoresList.scores), len: length)
+
+  var scoresSeq = newSeq[PDScore](length)
+  for i in 0 ..< length:
+    let score = cArray[i]
+    scoresSeq[i] = PDScore(value: score.value.uint32, rank: score.rank.uint32, player: $score.player)
+
+  for i in 0 ..< length:
+    playdate.scoreboards.freeScore(addr cArray[i])
+  cArray.data = nil
+
+  let domainObject = newPDScoresList(boardID = $scoresList.boardID, lastUpdated = scoresList.lastUpdated, scores = scoresSeq)
+  callback(domainObject, $errorMessage)
+  # todo enabling this cleanup code freezes the simulator. In this correct to free individual score first, right?
+  # playdate.scoreboards.freeScoresList(scoresList)
```

Comment:
How could I fix this? This code compiles but freezes when run.

Without it, the behaviour is as expected, but we have an obvious memory leak

Reply:
Fixed; turns out calling freeScoresList is the only thing that's needed